### DISCUSSION TITLE
userdefinedAction is trying to retrieve an optional parameter "lumping"

### DISCUSSION THREAD

**Discussion by abdurrahmanozturk** [idaholab/moose#21619] [Q&A General] [answered]
Hi,

I am trying to create an action that will set up a combination of TimeDerivative, MatDiffusion, MatReaction and MaskBodyForce kernels. I don't get any error during compilation, but when I try to run a test input file, it gives the following error.

"The parameter "lumping" is being retrieved before being set.  ([Line 1587 @ InputParameters.h](https://github.com/idaholab/moose/blob/b7a3ce3f6ca85f9f36873873de68db9d200bc44d/framework/include/utils/InputParameters.h#L1587))

The only kernel having "lumping" as input parameter is [TimeDerivative](https://github.com/idaholab/moose/blob/c42b80999f9a1a7aaa09c96484fc8a760f992359/framework/src/kernels/TimeDerivative.C#L26). It seems lumping is not a required input parameter for TimeDerivative class and has a default value(false).

I tried to set it up "false" in myAction, but I am still getting the same error while I am running the input file. Do you have any idea?

    for (unsigned int i = 0; i < var_num; i++)
    {
      std::string var_name = var_name_prefix + Moose::stringify(i);
      InputParameters params = _factory.getValidParams("TimeDerivative");
      params.set<NonlinearVariableName>("variable") = var_name;
      params.set<bool>("implicit") = true;
      params.set<bool>("lumping") = false;
      params.applyParameters(parameters());
      std::string kernel_name = var_name + "_td";
      _problem->addKernel("TimeDerivative", kernel_name, params);
    }

Thanks,
Apo

**Comment by GiudGiud** [comment, marked as answer]
Hello

This is weird indeed, we should not be able to get to that error with lumping since it has a default.
In the backtrace it should say which kernel is getting to this error. 

I think what you did in the action might have broken the parameters. 
Do you need to do this ?
```
   params.applyParameters(parameters());
```
I've never used that to set parameters

Guillaume

**Reply by abdurrahmanozturk** [reply]
Hi Guillaume,

Thanks for your reply,

I am using it to apply common parameters of TimeDerivative kernel to the myAction_parameters. I tried commenting it out and excluding "lumping" and defining it explicitly. Both didn't work. 

Here are some references:
[TensorMechanicsAction.C](https://github.com/idaholab/moose/blob/8175f8cb0ba17eaa5473470393b2c535986729a3/modules/tensor_mechanics/src/actions/TensorMechanicsAction.C#L605)
[PolycrystalKernelAction.C](https://github.com/idaholab/moose/blob/06bc1f73f0d4dfb3d01ce9ba77d6bb97346e07eb/modules/phase_field/src/action/PolycrystalKernelAction.C#L94)

Here is my code:

    #include "myKernelAction.h"
    #include "Conversion.h"
    #include "Factory.h"
    #include "FEProblem.h"

    registerMooseAction("myApp", myKernelAction, "add_kernel");

    InputParameters
    myKernelAction::validParams()
    {
      InputParameters params = Action::validParams();
      params.addClassDescription("Set up kernels");
      params.addRequiredParam<unsigned int>("var_num",
                                            "Number of variables");
      params.addRequiredParam<std::string>("var_name_prefix",
                                           "Prefix for the variable name");
      return params;
    }

    myKernelAction::myKernelAction(const InputParameters & params)
      : Action(params),
        _var_num(getParam<unsigned int>("var_num")),
        _var_name_prefix(getParam<unsigned int>("var_name_prefix"))
    {
    }

    void
    myKernelAction::act()
    {
      for (unsigned int i = 0; i < _var_num; i++)
      {
        std::string var_name = _var_name_prefix + Moose::stringify(i);

        // Add TimeDerivative kernel
      {
        InputParameters params = _factory.getValidParams("TimeDerivative");
        params.set<NonlinearVariableName>("variable") = var_name;
        params.set<bool>("implicit") = true;
        params.applyParameters(parameters());                          // note: commenting out apply method didn't work
        //params.applyParameters(parameters(),{"lumping"});   // note: excluding "lumping" in apply method did't work
        //params.set<bool>("lumping") = false;

        std::string kernel_name = var_name + "_td";
        _problem->addKernel("TimeDerivative",kernel_name, params);
      }
      // Add MatDiffusion kernel
      // Add MaskedBodyForce kernel
      // Add MatReaction kernel
      }

Apo

**Reply by GiudGiud** [reply]
must be coming from somewhere else then.

Let's double check, let's run this in a debugger and add a breakpoint on the abort and print the backtrace.
See those instructions
https://mooseframework.inl.gov/application_development/debugging.html

**Reply by abdurrahmanozturk** [reply]
Here is all bt

(lldb) bt
\* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
 \* frame #0: 0x000000012c0c0f50 libmpi.12.dylib`MPI_Abort
    frame #1: 0x0000000131ee9ebd libmoose-dbg.0.dylib`moose::internal::mooseErrorRaw(msg="Stack frames: 25\n0: 0   libmesh_dbg.0.dylib                 0x0000000140edf578 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 648\n1: 1   libmoose-dbg.0.dylib                0x0000000131ee9d1f moose::internal::mooseErrorRaw(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 879\n2: 2   libradtor_test-dbg.0.dylib          0x0000000108714908 void mooseError<char const (&) [16], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [40]>(char const (&) [16], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [40]) + 184\n3: 3   libradtor_test-dbg.0.dylib          0x0000000108714820 bool const& InputParameters::getParamHelper<bool>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, InputParameters const"..., prefix="") at MooseError.C:83:3
    frame #2: 0x0000000108714908 libradtor_test-dbg.0.dylib`void mooseError<char const (&) [16], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [40]>(args=<no value available>, args="lumping", args=<no value available>) [16], std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, char const (&) [40]) at MooseError.h:258:3
    frame #3: 0x0000000108714820 libradtor_test-dbg.0.dylib`bool const& InputParameters::getParamHelper<bool>(name="lumping", pars=0x000000014f092e88, (null)=0x0000000000000000) at InputParameters.h:1585:5
    frame #4: 0x000000010993d8d5 libxfem-dbg.0.dylib`bool const& MooseObject::getParam<bool>(this=0x000000015005ec18, name="lumping") const at MooseObject.h:266:10
    frame #5: 0x0000000130538394 libmoose-dbg.0.dylib`TimeDerivative::TimeDerivative(this=0x000000015005ec18, parameters=0x000000014f092e88) at TimeDerivative.C:32:38
    frame #6: 0x00000001305385cd libmoose-dbg.0.dylib`TimeDerivative::TimeDerivative(this=0x000000015005ec18, parameters=0x000000014f092e88) at TimeDerivative.C:33:1
    frame #7: 0x00000001305e5c25 libmoose-dbg.0.dylib`std::__1::__shared_ptr_emplace<TimeDerivative, std::__1::allocator<TimeDerivative> >::__shared_ptr_emplace<InputParameters const&>(this=0x000000015005ec00, __a=allocator<TimeDerivative> @ 0x00000003040eb450, __args=0x000000014f092e88) at shared_ptr.h:295:37
    frame #8: 0x00000001305e58fd libmoose-dbg.0.dylib`std::__1::__shared_ptr_emplace<TimeDerivative, std::__1::allocator<TimeDerivative> >::__shared_ptr_emplace<InputParameters const&>(this=0x000000015005ec00, __a=allocator<TimeDerivative> @ 0x00000003040eb488, __args=0x000000014f092e88) at shared_ptr.h:289:5
    frame #9: 0x00000001305e57cf libmoose-dbg.0.dylib`std::__1::shared_ptr<TimeDerivative> std::__1::allocate_shared<TimeDerivative, std::__1::allocator<TimeDerivative>, InputParameters const&, void>(__a=0x00000003040eb560, __args=0x000000014f092e88) at shared_ptr.h:954:55
    frame #10: 0x00000001305e56db libmoose-dbg.0.dylib`std::__1::shared_ptr<TimeDerivative> std::__1::make_shared<TimeDerivative, InputParameters const&, void>(__args=0x000000014f092e88) at shared_ptr.h:963:12
    frame #11: 0x00000001305e5616 libmoose-dbg.0.dylib`std::__1::shared_ptr<MooseObject> Registry::build<TimeDerivative, MooseObject>(parameters=0x000000014f092e88) at Registry.h:234:12
    frame #12: 0x0000000131de56c7 libmoose-dbg.0.dylib`Factory::create(this=0x000000015002c448, obj_name="TimeDerivative", name="C0_source", parameters=0x00000003040ec6d8, tid=0, print_deprecated=false) at Factory.C:125:14
    frame #13: 0x00000001316d9360 libmoose-dbg.0.dylib`std::__1::shared_ptr<KernelBase> Factory::create<KernelBase>(this=0x000000015002c448, obj_name="TimeDerivative", name="C0_source", parameters=0x00000003040ec6d8, tid=0) at Factory.h:300:38
    frame #14: 0x0000000131627fa4 libmoose-dbg.0.dylib`NonlinearSystemBase::addKernel(this=0x000000015082e618, kernel_name="TimeDerivative", name="C0_source", parameters=0x00000003040ec6d8) at NonlinearSystemBase.C:446:18
    frame #15: 0x0000000130d117e9 libmoose-dbg.0.dylib`FEProblemBase::addKernel(this=0x000000014f838a18, kernel_name="TimeDerivative", name="C0_source", parameters=0x00000003040ec6d8) at FEProblemBase.C:2372:8
    frame #16: 0x0000000112353f5d libphase_field-dbg.0.dylib`RateTheoryKernelAction::act(this=0x000000014f7170e8) at RateTheoryKernelAction.C:55:15
    frame #17: 0x0000000131413802 libmoose-dbg.0.dylib`Action::timedAct(this=0x000000014f7170e8) at Action.C:91:3
    frame #18: 0x000000013141316e libmoose-dbg.0.dylib`ActionWarehouse::executeActionsWithAction(this=0x000000015002c160, task="add_kernel") at ActionWarehouse.C:384:19
    frame #19: 0x0000000131449ee9 libmoose-dbg.0.dylib`ActionWarehouse::executeAllActions(this=0x000000015002c160) at ActionWarehouse.C:344:5
    frame #20: 0x0000000131e7204d libmoose-dbg.0.dylib`MooseApp::runInputFile(this=0x000000015002ba18) at MooseApp.C:1040:21
    frame #21: 0x0000000131e6bf21 libmoose-dbg.0.dylib`MooseApp::run(this=0x000000015002ba18) at MooseApp.C:1400:5
    frame #22: 0x0000000100014245 radtor-dbg`main(argc=3, argv=0x00000003040eda58) at main.C:33:8
    frame #23: 0x000000020003b51e dyld`start + 462

**Reply by GiudGiud** [reply]
ok it's totally coming from TimeDerivative
I m still stumped as to why though
@grmnptr anything jumps to your mind?

**Reply by grmnptr** [reply]
My only clue at the moment is that maybe the action is called too early. I would register a task in the action which gets executed around the same time when other kernels are added.

Also can you try retrieving the lumping parameter from the param to check if it gives you the same error within the action before creating the kernel?

**Reply by grmnptr** [reply]
Oh never mind, you already have the task there:
so what I would do is something like this:

```
if (_current_task == "add_kernel")
{
  for (unsigned int i = 0; i < _var_num; i++)
  {
    std::string var_name = _var_name_prefix + Moose::stringify(i);

    // Add TimeDerivative kernel
    {
      InputParameters params = _factory.getValidParams("TimeDerivative");
      params.set<NonlinearVariableName>("variable") = var_name;
      params.set<bool>("implicit") = true;
      params.applyParameters(parameters());                          // note: commenting out apply method didn't work
      //params.applyParameters(parameters(),{"lumping"});   // note: excluding "lumping" in apply method did't work
      //params.set<bool>("lumping") = false;

      std::string kernel_name = var_name + "_td";
      _problem->addKernel("TimeDerivative",kernel_name, params);
    }
    // Add MatDiffusion kernel
    // Add MaskedBodyForce kernel
    // Add MatReaction kernel
  }
```

**Reply by grmnptr** [reply]
And try to make sure that the task in your action depends on the regular add_kernel task from moose. You can do it like this in your app source file:

```
static void
associateSyntaxInner(Syntax & syntax, ActionFactory & /*action_factory*/)
{
  ...
  ...
  // add Kernels action
  registerTask("add_myapp_kernels", /*is_required=*/false);
  addTaskDependency("add_myapp_kernel", "add_kernel");
  ...
  ...
}
```

which assumes that you change this line in your action source file:

```
registerMooseAction("myApp", myKernelAction, "add_kernel");
```

to

```
registerMooseAction("myApp", myKernelAction, "add_myapp_kernel");
```

**Reply by abdurrahmanozturk** [reply]
It worked! 👍🏻

Thanks for help Peter. Now I can add other kernels.

Here is the summary of changes for who might have the same problem,

myKernelAction.C

    registerMooseAction("myApp", myKernelAction, "add_myapp_kernel");

    if (_current_task == "add_myapp_kernel")
    {
        for (unsigned int i = 0; i < _var_num; i++)
        {
             std::string var_name = _var_name_prefix + Moose::stringify(i);

            // Add TimeDerivative kernel
            {
              InputParameters params = _factory.getValidParams("TimeDerivative");
              params.set<NonlinearVariableName>("variable") = var_name;
              params.set<bool>("implicit") = true;
              params.applyParameters(parameters());                         
              std::string kernel_name = var_name + "_td";
              _problem->addKernel("TimeDerivative",kernel_name, params);
            }
        }
    }

myApp.C

        static void
        associateSyntaxInner(Syntax & syntax, ActionFactory & /*action_factory*/)
        {
        ...
        ...
        // add Kernels action
         registerTask("add_myapp_kernel", /*is_required=*/false);
        addTaskDependency("add_myapp_kernel", "add_kernel");
        ...
        ...
        }

Apo
